dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/orm/EntityParser.java

```diff
@@ -0,0 +1,747 @@
+                        superclass = superclass.getSuperclass()) {
+            parseAnnotatedObject(superclass);
+        }
+
+        return annotatedEntity;
+    }
+
+    /**
+     * Parses an unannotated generated entity from a record and constructs an
+     * orm model for that entity.
+     *
+     * @param record          the record class
+     * @param generatedEntity the generated entity class
+     * @return Class the record class
+     */
+    public Class<?> parseRecord(Class<?> record, Class<?> generatedEntity) {
+        if (doneParsing) {
+            // Internal exception
+            throw new IllegalStateException("Attempted to parse an entity after generating mapping");
+        }
+
+        entityToRecord.put(generatedEntity, record);
+        parse(generatedEntity, tablePrefix + record.getSimpleName());
+
+        return record;
+    }
+
+    /**
+     * Parses an unannotated entities found on a Repository interface and
+     * constructs an orm model for that entity.
+     *
+     * @param entity the entity class
+     * @return class the entity class
+     */
+    public Class<?> parseUnannotatedEntity(Class<?> entity) {
+        if (doneParsing) {
+            // Internal exception
+            throw new IllegalStateException("Attempted to parse an entity after generating mapping");
+        }
+
+        parse(entity, tablePrefix + entity.getSimpleName());
+
+        return entity;
+    }
+
+    // PARSER
+
+    /**
+     * Parses the entity and creates the orm of this entity.
+     *
+     * @param entity    the entity class
+     * @param tableName the table name, including prefix
+     */
+    @Trivial
+    private void parse(Class<?> entity, String tableName) {
+        this.currentEntity = entity;
+        this.idAttributes = new HashSet<>();
+
+        for (Class<?> superclass = currentEntity; //
+                        superclass != null && superclass != Object.class; //
+                        superclass = superclass.getSuperclass()) {
+
+            // Find annotated converters
+            for (Convert convert : superclass.getAnnotationsByType(Convert.class)) {
+                foundConverter(convert);
+            }
+
+            // Record entity and any embeddables found along the way
+            if (superclass == currentEntity && !entities.containsKey(superclass)) {
+                entities.put(superclass, new EntityRecord(superclass, tableName, finalizeAttributes(superclass, findAttributes(superclass))));
+                continue;
+            }
+
+            // Record entity-mappedSuperclass relationship
+            final Class<?> sc = superclass;
+            entitiesSuperclasses.computeIfPresent(currentEntity, (key, set) -> {
+                set.add(sc);
+                return set;
+            });
+            entitiesSuperclasses.computeIfAbsent(currentEntity, (key) -> {
+                return new HashSet<>(List.of(sc));
+            });
```

Comment:
This seemed a little confusing, and also less efficient. I think the goal is to do:
```suggestion
            Set<Class<?>> superclasses = entitiesSuperclasses.get(currentEntity);
            if (superclasses == null)
                entitiesSuperclasses.put(superclasses = new HashSet<>());
            superclasses.add(superclass);
```